wgpu-core/src/any_surface.rs

```diff
@@ -1,92 +1,93 @@
 use wgt::Backend;
 
-/// The `AnySurface` type: a `Arc` of a `HalSurface<A>` for any backend `A`.
+/// The `AnySurface` type: a `Arc` of a `A::Surface` for any backend `A`.
 use crate::hal_api::HalApi;
-use crate::instance::HalSurface;
 
-use std::any::Any;
 use std::fmt;
-use std::sync::Arc;
+use std::mem::ManuallyDrop;
+use std::ptr::NonNull;
 
-/// A `Arc` of a `HalSurface<A>`, for any backend `A`.
+struct AnySurfaceVtable {
+    // We oppurtunistically store the backend here, since we now it will be used
+    // with backend selection and it can be stored in static memory.
+    backend: Backend,
+    // Drop glue which knows how to drop the stored data.
+    drop: unsafe fn(*mut ()),
+}
+
+/// An `A::Surface`, for any backend `A`.
 ///
-/// Any `AnySurface` is just like an `Arc<HalSurface<A>>`, except that the
-/// `A` type parameter is erased. To access the `Surface`, you must
-/// downcast to a particular backend with the \[`downcast_ref`\] or
-/// \[`take`\] methods.
-pub struct AnySurface(Arc<dyn Any + 'static>);
+/// Any `AnySurface` is just like an `A::Surface`, except that the `A` type
+/// parameter is erased. To access the `Surface`, you must downcast to a
+/// particular backend with the \[`downcast_ref`\] or \[`take`\] methods.
+pub struct AnySurface {
+    data: NonNull<()>,
```

Comment:
> NonNull can be a footgun WRT subtyping

Covariance should only matter for a concrete `T` and particularly when it [leaks through as a type parameter](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=00d0f335799637af6dc7b13d389288be). The only property we're using here is niche filling (cannot inhabit zero bits). I don't mind changing it either way though.

> () is a ZST, and definitely not what's behind this pointer

Every raw thin pointer (but not unsized pointers) has the same layout and representation both from a Rust and an FFI perspective. The type parameter should only matter when it's being dereferenced, for which we are careful to cast it to the correct underlying type.

Regarding FFI, this is to my understanding currently just a private implementation detail. To make the whole structure FFI safe to the extend that we'd allow callers to access impleemntation details of it (e.g. assume that the first field is a pointer), we'd have to mark it `repr(C)`. If we do so, NonNull has been fine to use in FFI since 1.29 (when it received `#[repr(transparent)]`). Today it's guaranteed from a layout perspective to be exactly the same as `*mut T`.

I'd like it if you elaborated on how this is currently used in FFI, and how you want to use it.

Reply:
To make my position clear: I don't believe the properties of `NonNull` and `()` that I've brought up are actually problematic in this concrete usage. I'm mostly concerned with avoiding habits that can be problematic in other situations where APIs are being built on top of pointer-casting code. I doubt this code will change to ever expose `data`'s type parameter, so I don't see a latent long-term problem, either.

To me, what I bring up is a question of taste. I feel this conversation has been given due consideration from you as the PR author, so I don't consider the points I've raised to be blocking now. I'll invite @cwfitzgerald and @jimblandy to provide their thoughts, if any, but I'm not invested in a specific outcome.

RE: FFI usage: I think I used the term "FFI" more loosely than I should have, sorry about that. Concerns when authoring FFI overlap with my thinking here, to be sure. However, FFI is only tangentially related to this conversation because we use FFI to interface with the platforms that back surfaces.

The only usage of this specific API that I'm aware of is to consolidate a family of types into one (`AnySurface`) for the sake of an easier-to-write `wgpu_core::instance::Surface` API. In fact, I'm not even sure that exposing this as `pub` to consumers of `wgpu-core` is intended, since I don't see a way for `AnySurface`s to be used in our API at all (CC @cwfitzgerald, @jimblandy; is this right?).